service/src/PlatformTopo.cpp

```diff
@@ -55,9 +63,44 @@ int geopm_read_cpuid(void)
     return ((family << 8) + model);
 }
 
+static volatile unsigned g_is_popen_complete = 0;
+static struct sigaction g_popen_complete_signal_action;
+
+static void geopm_topo_popen_complete(int signum)
+{
+    if (signum == SIGCHLD) {
+        g_is_popen_complete = 1;
+    }
+}
+
+int geopm_topo_popen(const char *cmd, FILE **fid)
+{
+    int err = 0;
+    *fid = NULL;
+
+    struct sigaction save_action;
+    g_popen_complete_signal_action.sa_handler = geopm_topo_popen_complete;
+    sigemptyset(&g_popen_complete_signal_action.sa_mask);
+    g_popen_complete_signal_action.sa_flags = 0;
+    err = sigaction(SIGCHLD, &g_popen_complete_signal_action, &save_action);
+    if (!err) {
+        *fid = popen(cmd, "r");
+        while (*fid && !g_is_popen_complete) {
+
+        }
```

Comment:
It makes me a bit nervous to have this wait until the expected signal is received, especially because the interface this provides allows passing any string as the command for popen(). At the very least, input validation should be performed on cmd, but would also recommend establishing a limit on how long this loops.

This is more of a concern if this function is exposed more publicly. Otherwise, if it's only used internally and input to it is very well controlled, it's much less of a concern.

Reply:
I have moved the declaration of this out of the public header, so for the moment this is only used in the PlatformTopoImp, and the MSRIOGroupTest.  In both cases the input is very well controlled.

Moreover, this chunk of code was just moved around in this PR.  It is certainly a candidate for a refactor if we wanted to put this back into a public API.